validphys2/src/validphys/pseudodata.py

```diff
@@ -235,6 +242,402 @@ def indexed_make_replica(groups_index, make_replica):
+
+    [[CommonData(setname='NMC', ndata=204, commondataproc='DIS_NCE', nkin=3, nsys=16)]]
+    """
+    samples = [make_level1_data(level0_commondata_wc=level0_commondata_wc,
+                                    filterseed=filterseed+i,
+                                    data_index=data_index) for i in range(n_samples)]
+
+    return samples
+
+
+def sm_predictions(
+        dataset_inputs,
+        pdf,
+        theoryid
+    ):
+
+    """
+    Parameters
+    ----------
+    dataset_inputs: NSList of core.DataSetInput objects
+
+    pdf: core.PDF object
+
+    theoryid: TheoryIDSpec
+
+    Returns
+    -------
+
+    dict
+        dictionary of standard model predictions for the
+        given dataset_input, pdf, and theory
+
+    """
+    
+    sm_dict = {}
+
+    for dataset in dataset_inputs:
+        data = l.check_dataset(dataset.name, cfac=dataset.cfac, theoryid=theoryid)
+
+        sm_dict[dataset.name] = central_predictions(data, pdf)
+
+    return sm_dict
+
+
+def load_contamination(
+        contamination_parameters,
+        theoryid,
+        dataset_inputs
+    ):
+
+    """
+    Parameters
+    ----------
+
+    contamination_parameters: dict with 
+
+    theoryid: TheoryIDSpec
+
+    dataset_inputs: NSList of DataSetInput objects
+
+    Returns
+    -------
+
+    dict
+        dictionary of BSM k-factors to apply on certain datasets
+    
+    """
+
+    cont_path = l.datapath / f"theory_{theoryid.id}" / "simu_factors"
+
+    cont_name = contamination_parameters["name"]
+    cont_value = contamination_parameters["value"]
+    cont_lin_comb = contamination_parameters["linear_combination"]
+
+    bsm_dict = {}
+
+    for dataset in dataset_inputs:
+        
+        bsmfile = cont_path / f"SIMU_{dataset.name}.yaml"
+
+        cont_order = dataset.contamination
+
+        if cont_order == None:
+            log.warning(
+                f"{dataset.name} is not contaminated. Is it right?"
+            )
+            bsm_dict[dataset.name] = np.array([1.])
+        elif not os.path.exists(bsmfile):
+            log.error(
+                f"Could not find a BSM-factor for {dataset.name}. Are you sure they exist in the given theory?"
+            )
+            bsm_dict[dataset.name] = np.array([1.])
+        else:
+            log.info(
+                f"Loading {dataset.name}"
+            )
+            with open(bsmfile, "r+") as stream:
+                simu_card = yaml.safe_load(stream)
+            stream.close()
+
+            k_factors = np.zeros(len(simu_card["SM_fixed"]))
+            for op in cont_lin_comb:
+                k_factors += cont_lin_comb[op] * np.array(simu_card[cont_order][op])
+            k_factors = 1. + k_factors * cont_value / np.array(simu_card[cont_order]["SM"])
+
+            bsm_dict[dataset.name] = k_factors
+
+    return bsm_dict
+
+
+def compute_chi2(
+        make_level1_list_data,
+        sm_predictions,
+        groups_covmat,
+        load_contamination
```

Comment:
Also this should be moved somewhere else (`simunet_analysis.py`) and have another name that is more informative, given that this function can only be used for very specific cases and is not used to compute the chi2 in general